app/flatpak-builtins-build.c

```diff
@@ -75,6 +75,94 @@ child_setup (gpointer user_data)
     fcntl (g_array_index (fd_array, int, i), F_SETFD, 0);
 }
 
+static gboolean
+find_matching_extension_group_in_metakey (GKeyFile    *metakey,
+                                          const char  *id,
+                                          const char  *specified_tag,
+                                          char       **out_extension_group,
+                                          GError     **error)
+{
+  g_auto(GStrv) groups = NULL;
+  g_autofree char *extension_prefix = NULL;
+  const char *untagged_extension_group = NULL;
+  const char *last_seen_group = NULL;
+  guint n_extension_groups = 0;
+  GStrv iter = NULL;
+
+  g_return_val_if_fail (out_extension_group != NULL, FALSE);
+
+  groups =  g_key_file_get_groups (metakey, NULL);
+  extension_prefix = g_strconcat (FLATPAK_METADATA_GROUP_PREFIX_EXTENSION,
+                                  id,
+                                  NULL);
+
+  for (iter = groups; *iter != NULL; ++iter)
+    {
+      const char *group_name = *iter;
+      const char *extension_name = group_name + strlen (FLATPAK_METADATA_GROUP_PREFIX_EXTENSION);
+      g_autofree char *extension_tag = NULL;
+
+      flatpak_parse_extension_with_tag (extension_name,
+                                        NULL,
+                                        &extension_tag);
+
+      if (g_str_has_prefix (group_name, extension_prefix))
+        {
+          ++n_extension_groups;
+
+          /* Check 1: Does this extension have the same tag as the
+           * specified tag? If so, use it */
+          if (specified_tag != NULL &&
+              extension_tag != NULL &&
+              g_strcmp0 (extension_tag, specified_tag) == 0)
+            {
+              *out_extension_group = g_strdup (group_name);
+              return TRUE;
+            }
+
+          /* Check 2: Does this extension have no tag? If so, keep track of it */
+          if (extension_tag == NULL)
+            untagged_extension_group = group_name;
+
+          /* Check 3: Keep track of this extension group as the last
+           * seen one. If it was the only one then we can use it. */
+          last_seen_group = group_name;
+        }
+    }
+
+    if (n_extension_groups > 1)
+      {
+        if (untagged_extension_group != NULL)
```

Comment:
I wonder if this is right. Should we not also require specified_tag to be NULL here?
I.e. if you have multiple extensions  "foo" and "foo@2", should "foo@3" really match "foo"?

I think a better approach would be to exit out the loop immediately on both exact tag match, and if both specified and extension tag are unset. Then after the loop only handle the n_extension_groups == 1 case, and fail in all others.

Reply:
Makes sense.